src/Features/CSharp/Portable/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    internal partial class CSharpUseRangeOperatorDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public const string UseIndexer = nameof(UseIndexer);
+        public const string ComputedRange = nameof(ComputedRange);
+        public const string ConstantRange = nameof(ConstantRange);
+
+        public CSharpUseRangeOperatorDiagnosticAnalyzer() 
+            : base(IDEDiagnosticIds.UseRangeOperatorDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_range_operator), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources._0_can_be_simplified), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // We're going to be checking every invocation in the compilation. Cache information
+                // we compute in this object so we don't have to continually recompute it.
+                var infoCache = new InfoCache(compilationContext.Compilation);
+                compilationContext.RegisterOperationAction(
+                    c => AnalyzeInvocation(c, infoCache),
+                    OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(
+            OperationAnalysisContext context, InfoCache infoCache)
+        {
+            var cancellationToken = context.CancellationToken;
+            var invocation = (IInvocationOperation)context.Operation;
+
+            var invocationSyntax = invocation.Syntax as InvocationExpressionSyntax;
+            if (invocationSyntax is null ||
+                invocationSyntax.ArgumentList is null)
+            {
+                return;
+            }
+
+            // Check if we're at least on C# 8, and that the user wants these operators.
+            var syntaxTree = invocationSyntax.SyntaxTree;
+            var parseOptions = (CSharpParseOptions)syntaxTree.Options;
+            if (parseOptions.LanguageVersion < LanguageVersion.CSharp8)
+            {
+                return;
+            }
+
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet is null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CSharpCodeStyleOptions.PreferRangeOperator);
+            if (!option.Value)
+            {
+                return;
+            }
+
+
+            // look for `s.Slice(e1, end - e2)`
+            if (invocation.Instance is null ||
+                invocation.Instance.Syntax is null ||
```

Comment:
>invocation.Instance.Syntax is null [](start = 16, length = 34)

Nit: Redundant check, we [assert](http://source.roslyn.io/#Roslyn.Test.Utilities/Compilation/TestOperationVisitor.cs,48) non-null Syntax for every IOperation node in all our operation tests

Reply:
great!   i wasn't sure if that was an invariant or not.  Good to know!